src/Wrapper/KmsClientFactory.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Keboola\ObjectEncryptor\Wrapper;
+
+use Aws\Credentials\CredentialProvider;
+use Aws\Kms\KmsClient;
+use Aws\Sts\StsClient;
+use Keboola\ObjectEncryptor\EncryptorOptions;
+use Keboola\ObjectEncryptor\Exception\ApplicationException;
+
+class KmsClientFactory
+{
+    private const CONNECT_TIMEOUT = 10;
+    private const CONNECT_RETRIES = 5;
+    private const TRANSFER_TIMEOUT = 120;
+
+    public function createClient(EncryptorOptions $encryptorOptions): KmsClient
+    {
+        $region = $encryptorOptions->getKmsKeyRegion();
+        if ($region === null) {
+            throw new ApplicationException('Cipher key settings are missing.');
+        }
+
+        $role = $encryptorOptions->getKmsRole();
+
+        $stsClient = new StsClient([
+            'region' => $region,
+            'version' => '2011-06-15',
+            'retries' => self::CONNECT_RETRIES,
+            'http' => [
+                'connect_timeout' => self::CONNECT_TIMEOUT,
+                'timeout' => self::TRANSFER_TIMEOUT,
+            ],
+            'credentials' => false,
+        ]);
+
+        if ($role) {
+            $credentials = CredentialProvider::memoize(
+                CredentialProvider::assumeRole([
+                    'client' => $stsClient,
+                    'assume_role_params' => [
+                        'RoleArn' => $role,
+                        'RoleSessionName' => 'Encrypt-Decrypt',
+                    ],
+                ]),
+            );
```

Comment:
Cely by to mel byt jen poprehazovany kod toho, co tu bylo pred tim. Tohle je jediny kus, ktery jsme zmenil, aby bylo zajistene, ze pokud credentials expiruji, tak se automaticky refreshnou (predtim s eproste zavolal `assumeRole` a neresil se refresh, jelikoz cely klient se vytvarel porad dokola). Jeste budu muset vymyslet, jak to otestovat (pouziva se vubec explicitne zadana role?)

Reply:
A co že je teda změný ?

Reply:
Tak nakonec je tam test, ktery to pokryva a mam ozkouseny, ze to i takle ponovu, prohany skrz `memoize` funguje. Dokonce se mi podarilo najit v dokumentaci, ze to delaji stejne https://docs.aws.amazon.com/sdk-for-php/v3/developer-guide/guide_credentials_provider.html, takze to povazuju za cajk.